subsys/net/ip/net_pkt.c

```diff
@@ -888,6 +888,23 @@ void net_pkt_get_info(struct k_mem_slab **pkts,
 	}
 }
 
+struct net_pkt *net_pkt_pullup(struct net_pkt *pkt, int len)
+{
+	net_pkt_iter_init(pkt);
+
+	if (net_pkt_is_contiguous(pkt, len) == false) {
+
+		if (net_pkt_skip(pkt, len) < 0) {
```

Comment:
That confuses me. You should not be skipping the length you've been checking for contiguity here.

Well, maybe I misunderstood pullup but shouldn't this be used like (let's say for udp hdr):

- struct net_udp_hdr *hdr;
- net_pkt_skip(pkt, net_pkt_ip_hdr_len(pkt) + net_pkt_ipv6_ext_len(pkt)); <-- iter is at right position now
- if (net_pkt_pullup(pkt, NET_UDPH_LEN)) { manage error case, I assumed it returns a int here, so 0 on success }
- hdr= net_pkt_tod(pkt, struct net_udp_hdr *);

No?

(well in this case in means then pullup would be just an alias to net_pkt_is_contiguous() or we could just rename that one to pullup?)

Reply:
To me it looks like pullup should call skip, and shouldn't we call it like this anyway if one wants all protocol headers to be continuous.

net_pkt_pullup(pkt, net_pkt_ip_hdr_len(pkt) + net_pkt_ipv6_ext_len(pkt) + NET_UDPH_LEN);

Shouldn't we also "linearize" the the said memory area in pullup() and set the net_buf chain correctly?

Reply:
Ah yes, I got the idea of the skip in pullup now.
So initializing the iter are still wrong, but indeed in case of writing you want to skip the relevant length, but it needs to go backwark again at the position right before the skip:

So it requires a backup of the iter, and in case of success: it should restore that backup.

Reply:
> So it requires a backup of the iter, and in case of success: it should restore that backup.

OK, I thought about it. No, it's correct, let's please keep it like this.

The user is supposed to trim the previous protocols headers with net_pkt_adj(-x).

This actually simplifies things in typical use.

Reply:
Huh, trimming would kill completely the scheme I propose. You are not supposed to change the buffer data. Or then you'll need to store the original position in net_pkt.

Reply:
Please keep it as is. Users of this API don't deal with iterators. Look from the usage case.

If you would like to have a version with the iterator backup, just add such a version, but keep this one.

Reply:
I meant a backup/restor of the iterator in your function: right after net_pkt_is_contiguous(), on success. You backup, you skip: on success you restore. Then you know that pkt->iter.pos is pointing on a valid contiguous space.

Without that you won't get net_pkt_tod() working. Or it will point on the space "after" the contiguous space you checked.

Reply:
@tbursztyka 
> I meant a backup/restor of the iterator in your function: right after net_pkt_is_contiguous(), on success. You backup, you skip: on success you restore. Then you know that pkt->iter.pos is pointing on a valid contiguous space.
> Without that you won't get net_pkt_tod() working. Or it will point on the space "after" the contiguous space you checked.

OK. I am not sure that I understand you here 100%.

net_pkt_tod() is meant to be used only right after net_pkt_pullup(). I can still update the documentation to emphasize it, the wording might not be strong enough.

Otherwise, I am actually pretty sure it works as it should.

I absolutely don't mean to be offensive here, but, in case there are still doubts, can you please compile the unit test and run it under the debugger?

There are all paths explored in the unit test, happy and all erroneous.

Of course, I might be terribly missing something, I can't exlclude it.

Reply:
Well, if I understood net_pkt_tod() it is supposed to give you back a pointer on a contiguous area. 
usage would be:
- struct net_udp_hdr *hdr;
- net_pkt_pullup(pkt, sizeof(*hdr));
- hdr = net_pkt_tod(pkt)

Right?

If so, with current net_pkt_pullup() that's not going to work, since ptk->iter.pos (the one used by net_pkt_tod()) because this pointer is then pointing on the space after the one you want (due to net_pkt_skip()).